AppCenterIdentity/AppCenterIdentity/Internals/Storage/MSKeychainAuthTokenStorage.m

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import "MSKeychainAuthTokenStorage.h"
+#import "MSAuthTokenInfo.h"
+#import "MSIdentityConstants.h"
+#import "MSIdentityPrivate.h"
+#import "MSKeychainUtil.h"
+#import "MSLogger.h"
+#import "MSUtility.h"
+
+@implementation MSKeychainAuthTokenStorage
+
+@synthesize authTokensHistoryState = _authTokensHistoryState;
+
+- (nullable NSString *)retrieveAuthToken {
+  NSMutableArray<MSAuthTokenInfo *> *authTokensHistory = [self authTokensHistoryState];
+  MSAuthTokenInfo *latestAuthTokenInfo = authTokensHistory.lastObject;
+  return latestAuthTokenInfo.authToken;
+}
+
+- (nullable NSString *)retrieveAccountId {
+  return [MS_USER_DEFAULTS objectForKey:kMSIdentityMSALAccountHomeAccountKey];
+}
+
+// TODO: Finish the implementation, this method will be used to get logs from DB for a period when token was active.
+- (MSAuthTokenInfo *)oldestAuthToken {
+  return [MSAuthTokenInfo new];
+
+  /*
+  // Read token array from storage.
+  NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+  if (!tokenArray || tokenArray.count == 0) {
+    return nil;
+  }
+
+  MSAuthTokenInfo *authTokenInfo = tokenArray.firstObject;
+  NSDate *nextChangeTime = tokenArray.count > 1 ? tokenArray[1].startTime : nil;
+  if ([authTokenInfo.endTime laterDate:nextChangeTime]) {
+    return [[MSAuthTokenInfo alloc] initWithAuthToken:authTokenInfo.authToken
+                                         andStartTime:authTokenInfo.startTime
+                                           andEndTime:nextChangeTime];
+  }
+  return authTokenInfo;
+   */
+}
+
+- (void)saveAuthToken:(nullable NSString *)authToken withAccountId:(nullable NSString *)accountId expiresOn:(nullable NSDate *)expiresOn {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *authTokensHistory = [self authTokensHistoryState];
+    if (authTokensHistory.count == 0) {
+
+      /*
+       * Adding a nil entry is required during the first initialization to differentiate
```

Comment:
If we need to add a `nil` entry to history, it can be done by core (like session history). I still don't see any reason of doing this one because anonymous usage could be covered if it isn't in the history. Am I missing anything here?

Reply:
Disagree.

> because anonymous usage could be covered if it isn't in the history

No, it can't.

> Am I missing anything here?

Probably the history size limit. We shouldn't send logs related to deleted token as anonymous.

Reply:
Well. I still think `nil` should be added at the time the service starts. This is a valid data in between the SDK start and sign-in especially for crashes.
BTW, are we dropping logs that cannot be found an identity for the logs? I think we still send it as anonymous. Please double check.

Reply:
Disagree again. The key point here is the start time. In your case logs from previous runs/older SDKs version are out of range. This check fixes it.

Reply:
Regarding dropping logs, it's in the separate ongoing PR

Reply:
I assume it drops logs if it cannot find identity from history, this implementation will drop some of anonymous logs.
Here is a scenario.
1. App launched at 1:00 PM
2. App Crashed at 1:01 PM
3. App launched again at 1:02 PM, at this time previous crashes will be processed. Assume first and second tries failed and it is waiting for the last retry.
4. User sign-in at 1:05 PM
5. App Crashed at 1:06 PM. At this time the history will contain "nil/anonymous" at 1:05 PM and one user identity at 1:05 PM
6. App relaunched at 1:07 PM

The first crash happened at 1:01 PM and it couldn't find any identities in the history so the log will be dropped.

If you add "nil/anonymous" identity to the history at the time Identity starts, the history will contain "nil/anonymous" with the time of 1:00 PM so it will use this anonymous identity (not dropping logs).

Reply:
And putting an anonymous identity when history is empty isn't necessary anymore.

Reply:
Anyway dropping logs is not a part of this PR. It's not a blocker. It will be fixed in next one.